src/scheduler/src/main/java/org/cloudfoundry/autoscaler/scheduler/rest/BasicAuthenticationFilter.java

```diff
@@ -0,0 +1,206 @@
+import org.springframework.core.annotation.Order;
+import org.springframework.http.HttpHeaders;
+import org.springframework.stereotype.Component;
+import org.springframework.util.ObjectUtils;
+
+@Slf4j
+@Component
+@Order(2)
+public class BasicAuthenticationFilter implements Filter {
+  private static final String WWW_AUTHENTICATE_VALUE = "Basic";
+  private static final Map<String, Boolean> validProtectedEndpointsMap;
+
+  static {
+    validProtectedEndpointsMap =
+        Map.of(
+            "/health/prometheus", true,
+            "/health/liveness", true);
+  }
+
+  final HealthServerConfiguration healthServerConfiguration;
+
+  public BasicAuthenticationFilter(HealthServerConfiguration healthServerConfiguration) {
+    this.healthServerConfiguration = healthServerConfiguration;
+  }
+
+  @Override
+  public void doFilter(
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+      throws IOException, ServletException {
+    HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
+    HttpServletResponse httpResponse = (HttpServletResponse) servletResponse;
+    List<String> unprotectedEndpointsConfig = healthServerConfiguration.getUnprotectedEndpoints();
+
+    if (!httpRequest.getRequestURI().contains("/health")) {
+      log.debug("Not a health request: " + httpRequest.getRequestURI());
+      chain.doFilter(servletRequest, servletResponse);
+      return;
+    }
+
+    if (ObjectUtils.isEmpty(unprotectedEndpointsConfig)) { // health endpoints are authorized
+      isUserAuthenticatedOrSendError(chain, httpRequest, httpResponse);
+    } else if (!ObjectUtils.isEmpty(unprotectedEndpointsConfig)) {
+      Map<String, Boolean> validateMap = checkValidEndpoints(unprotectedEndpointsConfig);
+      if (!ObjectUtils.isEmpty(validateMap)) {
+        log.warn(
+            "Health configuration: invalid unprotectedEndpoints provided: "
+                + validateMap.get("invalidEndpoints"));
+        httpResponse.setHeader(HttpHeaders.WWW_AUTHENTICATE, WWW_AUTHENTICATE_VALUE);
+        httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+        return;
+      }
+      Map<String, Boolean> unprotectedConfig = getMapFromList(unprotectedEndpointsConfig);
+      List<String> allowedEndpointsWithoutBasicAuth =
+          areEndpointsAuthorized(unprotectedConfig, httpRequest.getRequestURI());
+
+      if (!allowedEndpointsWithoutBasicAuth.contains(httpRequest.getRequestURI())) {
+        log.error(
+            "Health configuration: Basic auth is required to access protectedEndpoints: "
+                + httpRequest.getRequestURI()
+                + " \nValid unprotected endpoints are: "
+                + allowedEndpointsWithoutBasicAuth);
+        httpResponse.setHeader("WWW-Authenticate", WWW_AUTHENTICATE_VALUE);
+        httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+        return;
+      }
+      chain.doFilter(servletRequest, servletResponse);
+    }
```

Comment:
The logic here looks broken. I wonder if it fails in the following scenario:
 * There is – by configuration – one protected endpoint "/health/prometheus" and one unprotected "/health/liveness".
 * The user is authenticated.
 * The user queries on "/health/prometheus".
 * Expected behaviour: The request will be handeled successfully.
 * Likely actual behaviour by this implementation: User gets blocked and receives an error message triggered by line 81.

Reply:
As the comments so far affect a lot of code and likely will cause the rest of this class to be heavily changed and shortened, I continue with other classes until this one is changed again.

Reply:
Nice finding... this is a bug and should be fixed